File: web/src/lib/tableDetection.ts
import { binarize, bridgeGaps, findRuledLines, longestRunPerColumn, longestRunPerRow, toGrayscale, type RasterImage } from './gridDetection'

export interface DetectedTable {
  rowLines: number[]
  colLines: number[]
  tableLeft: number
  tableRight: number
  tableTop: number
  tableBottom: number
}

export interface DetectTableOptions {
  xStart?: number
  xEnd?: number
  binarizeThreshold?: number
  gapBridgeRadius?: number
  rowLineMinFraction?: number
  colLineMinFraction?: number
  /**
   * Where the mostly-blank answer-column region starts, as a fraction of
   * crop width from xStart. Calibrated against the real sample form
   * (samples/Student_Questionnaire_BP_Pujari_1.pdf): the Sr.No/Code/Statement
   * block occupies the left ~21% of a half-page crop at this render scale.
   * This is a form-layout constant (PRODUCT.md confirms the layout is fixed
   * across print runs), not a general one — a different questionnaire would
   * need this recalibrated. An index-based approach (count coarse column
   * lines from the left) was tried first and was fragile: the table's own
   * outer-left border sometimes gets detected as a strong line too, throwing
   * off which ordinal position means "Statement column boundary."
   */
  blankRegionStartFraction?: number
}

const DEFAULTS = {
  binarizeThreshold: 210,
  gapBridgeRadius: 3,
  rowLineMinFraction: 0.5,
  colLineMinFraction: 0.35,
  blankRegionStartFraction: 0.21,
}

/**
 * Ruled-table detection, tuned against a real scanned page (see gridDetection.ts
 * for why gap-bridging is load-bearing — a naive threshold-and-count approach
 * found zero lines on the real sample). Two passes:
 *
 * 1. Find row lines restricted to the mostly-blank answer-column region
 *    (Sr.No/Code/Statement excluded via `blankRegionStartFraction`) — much
 *    cleaner signal than searching across the text-heavy Statement column,
 *    which produces text-line density peaks indistinguishable from ruled
 *    lines by a simple threshold.
 * 2. Find column lines restricted to the row span from step 1 — sharper
 *    than searching the full page height, now that the table's exact
 *    vertical extent is known.
 */
export function detectTable(image: RasterImage, options: DetectTableOptions = {}): DetectedTable {
  const opts = { ...DEFAULTS, ...options }
  const xStart = options.xStart ?? 0
  const xEnd = options.xEnd ?? image.width
  const empty: DetectedTable = { rowLines: [], colLines: [], tableLeft: xStart, tableRight: xEnd, tableTop: 0, tableBottom: 0 }

  const gray = toGrayscale(image)
  const binary = binarize(gray, opts.binarizeThreshold)
  const rowBridged = bridgeGaps(binary, image.width, image.height, 'row', opts.gapBridgeRadius)
  const colBridged = bridgeGaps(binary, image.width, image.height, 'column', opts.gapBridgeRadius)

  const blankRegionStart = Math.round(xStart + opts.blankRegionStartFraction * (xEnd - xStart))

  // Pass 1: row lines, searched only within the blank answer-column region.
  const rowSpan = xEnd - blankRegionStart
  const rowProfile = longestRunPerRow(rowBridged, image.width, image.height, blankRegionStart, xEnd)
  const rowLines = findRuledLines(rowProfile, rowSpan * opts.rowLineMinFraction)

  if (rowLines.length < 2) return { ...empty, rowLines }

  const tableTop = rowLines[0]
  const tableBottom = rowLines[rowLines.length - 1]

  // Pass 2: column lines, restricted to the known row span.
  const rowSpanHeight = tableBottom - tableTop
  const colProfile = longestRunPerColumn(colBridged, image.width, image.height, tableTop, tableBottom)
  const colLines = findRuledLines(colProfile, rowSpanHeight * opts.colLineMinFraction).filter((x) => x >= xStart && x < xEnd)

  return {
    rowLines,
    colLines,
    tableLeft: colLines[0] ?? xStart,
    tableRight: colLines[colLines.length - 1] ?? xEnd,
    tableTop,
    tableBottom,
  }
}

/** Draws the detected grid as red lines on top of the (already-drawn) canvas, for eyeballing. */
export function drawTableOverlay(ctx: CanvasRenderingContext2D, table: DetectedTable) {
  ctx.save()
  ctx.strokeStyle = 'red'
  ctx.lineWidth = 1
  for (const y of table.rowLines) {
    ctx.beginPath()
    ctx.moveTo(table.tableLeft, y)
    ctx.lineTo(table.tableRight, y)
    ctx.stroke()
  }
  for (const x of table.colLines) {
    ctx.beginPath()
    ctx.moveTo(x, table.tableTop)
    ctx.lineTo(x, table.tableBottom)
    ctx.stroke()
  }
  ctx.restore()
}

File: parser/lib/src/table_detection.dart
import 'package:image/image.dart' as img;

import 'grid_detection.dart';

class DetectedTable {
  final List<int> rowLines;
  final List<int> colLines;
  final int tableLeft;
  final int tableRight;
  final int tableTop;
  final int tableBottom;

  DetectedTable({
    required this.rowLines,
    required this.colLines,
    required this.tableLeft,
    required this.tableRight,
    required this.tableTop,
    required this.tableBottom,
  });

  static DetectedTable empty(int xStart, int xEnd) => DetectedTable(
        rowLines: [],
        colLines: [],
        tableLeft: xStart,
        tableRight: xEnd,
        tableTop: 0,
        tableBottom: 0,
      );
}

/// Ruled-table detection, tuned against the real sample scan
/// (samples/Student_Questionnaire_BP_Pujari_1.pdf) — ported directly from the
/// web track's `tableDetection.ts`, where this exact approach was measured
/// against the same file. See grid_detection.dart for why gap-bridging is
/// load-bearing here — a naive threshold-and-count approach found zero lines.
///
/// Two passes:
/// 1. Row lines, restricted to the mostly-blank answer-column region
///    ([blankRegionStartFraction] of crop width from xStart) — much cleaner
///    signal than searching across the text-heavy Statement column.
/// 2. Column lines, restricted to the row span from pass 1 — sharper than
///    searching the full page height.
DetectedTable detectTable(
  img.Image image, {
  int? xStart,
  int? xEnd,
  int binarizeThreshold = 210,
  int gapBridgeRadius = 3,
  double rowLineMinFraction = 0.5,
  double colLineMinFraction = 0.35,
  double blankRegionStartFraction = 0.21,
}) {
  final xs = xStart ?? 0;
  final xe = xEnd ?? image.width;

  final gray = toGrayscale(image);
  final binary = binarize(gray, binarizeThreshold);
  final rowBridged = bridgeGaps(binary, image.width, image.height, 'row', gapBridgeRadius);
  final colBridged = bridgeGaps(binary, image.width, image.height, 'column', gapBridgeRadius);

  final blankRegionStart = (xs + blankRegionStartFraction * (xe - xs)).round();

  final rowSpan = xe - blankRegionStart;
  final rowProfile = longestRunPerRow(rowBridged, image.width, image.height, xStart: blankRegionStart, xEnd: xe);
  final rowLines = findRuledLines(rowProfile, rowSpan * rowLineMinFraction);

  if (rowLines.length < 2) return DetectedTable.empty(xs, xe);

  final tableTop = rowLines.first;
  final tableBottom = rowLines.last;

  final rowSpanHeight = tableBottom - tableTop;
  final colProfile = longestRunPerColumn(colBridged, image.width, image.height, yStart: tableTop, yEnd: tableBottom);
  final colLines = findRuledLines(colProfile, rowSpanHeight * colLineMinFraction).where((x) => x >= xs && x < xe).toList();

  return DetectedTable(
    rowLines: rowLines,
    colLines: colLines,
    tableLeft: colLines.isNotEmpty ? colLines.first : xs,
    tableRight: colLines.isNotEmpty ? colLines.last : xe,
    tableTop: tableTop,
    tableBottom: tableBottom,
  );
}

/// Draws the detected grid as red lines on top of [image], for eyeballing.
void drawTableOverlay(img.Image image, DetectedTable table) {
  final red = img.ColorRgb8(255, 0, 0);
  for (final y in table.rowLines) {
    img.drawLine(image, x1: table.tableLeft, y1: y, x2: table.tableRight, y2: y, color: red);
  }
  for (final x in table.colLines) {
    img.drawLine(image, x1: x, y1: table.tableTop, x2: x, y2: table.tableBottom, color: red);
  }
}
